src/v/cloud_storage/topic_manifest_downloader.cc

```diff
@@ -109,4 +112,106 @@ topic_manifest_downloader::download_manifest(
+      hash_prefix);
+    retry_chain_node retry(deadline, backoff, &parent_retry);
+    auto prefixed_list_res = co_await remote.list_objects(
+      bucket,
+      retry,
+      cloud_storage_clients::object_key{fmt::format("{}/", hash_prefix)});
+    if (prefixed_list_res.has_error()) {
+        co_return error_outcome::manifest_download_error;
+    }
+    co_return prefixed_list_res.value().contents;
+}
+} // namespace
+
+ss::future<result<find_topic_manifest_outcome, error_outcome>>
+topic_manifest_downloader::find_manifests(
+  remote& remote,
+  cloud_storage_clients::bucket_name bucket,
+  retry_chain_node& parent_retry,
+  ss::lowres_clock::time_point deadline,
+  model::timestamp_clock::duration backoff,
+  std::optional<tp_ns_filter_t> tp_filter,
+  chunked_vector<topic_manifest>* manifests) {
+    retry_chain_node retry(deadline, backoff, &parent_retry);
+    absl::btree_set<model::topic_namespace> topics;
+    const auto maybe_add_topic =
+      [&topics,
+       &tp_filter](const std::optional<model::topic_namespace>& topic) {
+          if (!topic.has_value()) {
+              return;
+          }
+          if (!tp_filter.has_value() || tp_filter.value()(*topic)) {
+              topics.emplace(*topic);
+          }
+      };
+    {
+        // First, collect labeled topic manifests that match out filter.
+        const auto labeled_list_res = co_await remote.list_objects(
+          bucket,
+          retry,
+          cloud_storage_clients::object_key{
+            fmt::format("{}/", labeled_topic_manifests_root())});
+        if (labeled_list_res.has_error()) {
+            co_return error_outcome::manifest_download_error;
+        }
+        for (const auto& item : labeled_list_res.value().contents) {
+            auto tp_ns = from_labeled_topic_manifest_path(item.key);
+            maybe_add_topic(tp_ns);
+        }
+        // End scope to free memory.
+    }
+    // Next, collect prefixed topic manifests that match out filter.
+    const auto prefixed_roots = prefixed_topic_manifests_roots();
+    std::vector<ss::future<result<list_outcome_t, error_outcome>>> futs;
+    for (const auto& root : prefixed_roots) {
+        futs.push_back(find_prefixed_manifest_paths(
+          remote, bucket, parent_retry, deadline, backoff, root));
+    }
+    auto prefixed_res = co_await ss::when_all_succeed(futs.begin(), futs.end());
+    for (const auto& r : prefixed_res) {
+        if (r.has_error()) {
+            co_return error_outcome::manifest_download_error;
+        }
+        for (const auto& item : r.value()) {
+            auto tp_ns = from_prefixed_topic_manifest_path(item.key);
+            maybe_add_topic(tp_ns);
+        }
+    }
+    // Use the manifest downloader to look for the filtered manifests.
+    chunked_vector<topic_manifest> m;
+    m.reserve(topics.size());
+    for (const auto& tp : topics) {
+        topic_manifest tm;
+        topic_manifest_downloader dl(bucket, std::nullopt, tp, remote);
```

Comment:
nit: `/*hint=*/std::nullopt` here for readability

Reply:
Done